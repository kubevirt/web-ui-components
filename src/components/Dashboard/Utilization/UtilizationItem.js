import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Measure from 'react-measure';

import { Col, Row } from 'patternfly-react';
import { ChartArea, ChartVoronoiContainer, Chart, ChartAxis, ChartTooltip } from '@patternfly/react-charts';

import { MEDIA_QUERY_SM } from '../../../utils';

import { InlineLoading } from '../../Loading';
import { NOT_AVAILABLE } from './strings';

export class UtilizationItem extends React.PureComponent {
  state = {
    dimensions: {
      width: -1,
    },
  };

  onResize = contentRect => {
    this.setState({ dimensions: contentRect.bounds });
  };

  render() {
    const { id, title, data, maxY, unit, isLoading, LoadingComponent } = this.props;
    const { width } = this.state.dimensions;

    const axis = {
      y: {
        show: true,
        padding: {
          top: 0,
          bottom: 0,
        },
        tick: {
          count: 3,
          format: v => `${v} ${unit}`,
        },
      },
      x: {
        show: false,
      },
    };
    if (maxY) {
      axis.y.max = maxY;
    }

    let actual;
    let chart = NOT_AVAILABLE;
    if (isLoading) {
      chart = <LoadingComponent />;
    } else if (data) {
      const chartData = data.map((val, index) => ({ x: index, y: val ? Number(val.toFixed(1)) : 0 }));
      actual = `${Math.round(data[data.length - 1])} ${unit}`;
      chart = (
        <Chart
          id={id}
          height={60}
          width={300}
          containerComponent={
            <ChartVoronoiContainer
              labels={datum => `${datum.y} ${unit}`}
              labelComponent={<ChartTooltip style={{ fontSize: 12, padding: 5 }} />}
            />
          }
          padding={{ top: 10, left: 40, bottom: 5 }}
          domainPadding={{ x: 0, y: 10 }}
        >
          <ChartArea data={chartData} />
          <ChartAxis dependentAxis tickValues={[0, maxY / 2, maxY]} style={{ tickLabels: { fontSize: 10 } }} />
        </Chart>
      );
    }

    let topClass;
    let rows;
    if (width < MEDIA_QUERY_SM) {
      topClass = 'kubevirt-utilization__item-narrow';
      rows = (
        <Fragment>
          <Row>
            <Col lg={6} md={6} sm={6} xs={6} className="kubevirt-utilization__item-narrow-title">
              {title}
            </Col>
            <Col className="kubevirt-utilization__item-actual item-actual--narrow" lg={6} md={6} sm={6} xs={6}>
              {actual}
            </Col>
          </Row>
          <Row>
            <Col className="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--narrow">{chart}</Col>
          </Row>
        </Fragment>
      );
    } else {
      topClass = 'kubevirt-utilization__item';
      rows = (
        <Row>
          <Col lg={3} md={3} sm={3} xs={3}>
            {title}
          </Col>
          <Col className="kubevirt-utilization__item-actual" lg={2} md={2} sm={2} xs={2}>
            {actual}
          </Col>
          <Col
            className="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--wide"
            lg={7}
            md={7}
            sm={7}
            xs={7}
          >
            {chart}
          </Col>
        </Row>
      );
    }

    return (
      <Measure bounds onResize={this.onResize}>
        {({ measureRef }) => (
          <div ref={measureRef} className={topClass}>
            {rows}
          </div>
        )}
      </Measure>
    );
  }
}

UtilizationItem.defaultProps = {
  maxY: null,
  data: null,
  LoadingComponent: InlineLoading,
  isLoading: false,
};

UtilizationItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  data: PropTypes.array,
  unit: PropTypes.string.isRequired,
  maxY: PropTypes.number,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  isLoading: PropTypes.bool,
};
