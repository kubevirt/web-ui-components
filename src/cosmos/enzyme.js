import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
