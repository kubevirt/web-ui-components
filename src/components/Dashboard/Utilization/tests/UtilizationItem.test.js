import React from 'react';
import { render } from 'enzyme';

import { UtilizationItem } from '../UtilizationItem';
import { default as UtilizationItemFixtures } from '../fixtures/UtilizationItem.fixture';

// eslint-disable-next-line react/prop-types
const testItem = ({ props }) => <UtilizationItem {...props} />;

describe('<UtilizationItem />', () => {
  it('renders correctly', () => {
    UtilizationItemFixtures.forEach(fixture => {
      const component = render(testItem(fixture));
      expect(component).toMatchSnapshot();
    });
  });
});
