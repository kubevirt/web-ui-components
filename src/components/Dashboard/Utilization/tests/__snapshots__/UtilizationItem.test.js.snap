// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UtilizationItem /> renders correctly 1`] = `
<div
  class="kubevirt-utilization__item-narrow"
>
  <div
    class="row"
  >
    <div
      class="kubevirt-utilization__item-narrow-title col-lg-6 col-md-6 col-sm-6 col-xs-6"
    >
      title
    </div>
    <div
      class="kubevirt-utilization__item-actual item-actual--narrow col-lg-6 col-md-6 col-sm-6 col-xs-6"
    >
      5 %
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--narrow"
    >
      <div
        class="VictoryContainer"
        style="height:100%;width:100%;user-select:none;pointer-events:none;touch-action:none;position:relative"
      >
        <svg
          aria-labelledby="victory-container-1-title victory-container-1-desc"
          height="60"
          role="img"
          style="pointer-events:all;width:100%;height:100%"
          viewBox="0 0 300 60"
          width="300"
        >
          <g
            clip-path="url(#victory-clip-2)"
            style="height:100%;width:100%;user-select:none"
          >
            <defs>
              <clipPath
                id="victory-clip-2"
              >
                <rect
                  height="45"
                  role="presentation"
                  shape-rendering="auto"
                  vector-effect="non-scaling-stroke"
                  width="260"
                  x="40"
                  y="10"
                />
              </clipPath>
            </defs>
            <g>
              <path
                d="M40,51.46116504854369L105,55L170,49.69174757281554L235,19.611650485436897L300,53.230582524271846L300,55L235,55L170,55L105,55L40,55Z"
                role="presentation"
                shape-rendering="auto"
                style="fill:#bee1f4;fill-opacity:0.4;stroke:none;stroke-width:2"
              />
              <path
                d="M40,51.46116504854369L105,55L170,49.69174757281554L235,19.611650485436897L300,53.230582524271846"
                role="presentation"
                shape-rendering="auto"
                style="fill:none;fill-opacity:0.4;stroke:#39a5dc;stroke-width:2"
              />
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke:#72767b;fill:transparent;stroke-width:1;stroke-linecap:round;stroke-linejoin:round"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="40"
              y1="10"
              y2="55"
            />
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:none;fill:none;pointer-events:painted"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="300"
                y1="55"
                y2="55"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:transparent;fill:transparent;size:1px"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="39"
                y1="55"
                y2="55"
              />
              <text
                direction="inherit"
                dx="0"
                dy="3.55"
                x="29"
                y="55"
              >
                <tspan
                  dx="0"
                  style="font-size:10px;font-family:'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif;letter-spacing:normal;padding:10px;fill:#282d33;stroke:transparent"
                  text-anchor="end"
                  x="29"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:none;fill:none;pointer-events:painted"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="300"
                y1="55"
                y2="55"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:transparent;fill:transparent;size:1px"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="39"
                y1="55"
                y2="55"
              />
            </g>
          </g>
        </svg>
        <div
          style="z-index:99;position:absolute;top:0;left:0;width:100%;height:100%"
        >
          <svg
            height="60"
            style="overflow:visible;width:100%;height:100%"
            viewBox="0 0 300 60"
            width="300"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<UtilizationItem /> renders correctly 2`] = `
<div
  class="kubevirt-utilization__item-narrow"
>
  <div
    class="row"
  >
    <div
      class="kubevirt-utilization__item-narrow-title col-lg-6 col-md-6 col-sm-6 col-xs-6"
    >
      title
    </div>
    <div
      class="kubevirt-utilization__item-actual item-actual--narrow col-lg-6 col-md-6 col-sm-6 col-xs-6"
    >
      5 %
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--narrow"
    >
      <div
        class="VictoryContainer"
        style="height:100%;width:100%;user-select:none;pointer-events:none;touch-action:none;position:relative"
      >
        <svg
          aria-labelledby="victory-container-3-title victory-container-3-desc"
          height="60"
          role="img"
          style="pointer-events:all;width:100%;height:100%"
          viewBox="0 0 300 60"
          width="300"
        >
          <g
            clip-path="url(#victory-clip-4)"
            style="height:100%;width:100%;user-select:none"
          >
            <defs>
              <clipPath
                id="victory-clip-4"
              >
                <rect
                  height="45"
                  role="presentation"
                  shape-rendering="auto"
                  vector-effect="non-scaling-stroke"
                  width="260"
                  x="40"
                  y="10"
                />
              </clipPath>
            </defs>
            <g>
              <path
                d="M40,51.46116504854369L105,55L170,49.69174757281554L235,19.611650485436897L300,53.230582524271846L300,55L235,55L170,55L105,55L40,55Z"
                role="presentation"
                shape-rendering="auto"
                style="fill:#bee1f4;fill-opacity:0.4;stroke:none;stroke-width:2"
              />
              <path
                d="M40,51.46116504854369L105,55L170,49.69174757281554L235,19.611650485436897L300,53.230582524271846"
                role="presentation"
                shape-rendering="auto"
                style="fill:none;fill-opacity:0.4;stroke:#39a5dc;stroke-width:2"
              />
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke:#72767b;fill:transparent;stroke-width:1;stroke-linecap:round;stroke-linejoin:round"
              vector-effect="non-scaling-stroke"
              x1="40"
              x2="40"
              y1="10"
              y2="55"
            />
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:none;fill:none;pointer-events:painted"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="300"
                y1="55"
                y2="55"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:transparent;fill:transparent;size:1px"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="39"
                y1="55"
                y2="55"
              />
              <text
                direction="inherit"
                dx="0"
                dy="3.55"
                x="29"
                y="55"
              >
                <tspan
                  dx="0"
                  style="font-size:10px;font-family:'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif;letter-spacing:normal;padding:10px;fill:#282d33;stroke:transparent"
                  text-anchor="end"
                  x="29"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:none;fill:none;pointer-events:painted"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="300"
                y1="37.30582524271845"
                y2="37.30582524271845"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:transparent;fill:transparent;size:1px"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="39"
                y1="37.30582524271845"
                y2="37.30582524271845"
              />
              <text
                direction="inherit"
                dx="0"
                dy="3.55"
                x="29"
                y="37.30582524271845"
              >
                <tspan
                  dx="0"
                  style="font-size:10px;font-family:'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif;letter-spacing:normal;padding:10px;fill:#282d33;stroke:transparent"
                  text-anchor="end"
                  x="29"
                >
                  50
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:none;fill:none;pointer-events:painted"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="300"
                y1="19.611650485436897"
                y2="19.611650485436897"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke:transparent;fill:transparent;size:1px"
                vector-effect="non-scaling-stroke"
                x1="40"
                x2="39"
                y1="19.611650485436897"
                y2="19.611650485436897"
              />
              <text
                direction="inherit"
                dx="0"
                dy="3.55"
                x="29"
                y="19.611650485436897"
              >
                <tspan
                  dx="0"
                  style="font-size:10px;font-family:'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif;letter-spacing:normal;padding:10px;fill:#282d33;stroke:transparent"
                  text-anchor="end"
                  x="29"
                >
                  100
                </tspan>
              </text>
            </g>
          </g>
        </svg>
        <div
          style="z-index:99;position:absolute;top:0;left:0;width:100%;height:100%"
        >
          <svg
            height="60"
            style="overflow:visible;width:100%;height:100%"
            viewBox="0 0 300 60"
            width="300"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
